flink-clients/src/main/java/org/apache/flink/client/program/artifact/ArtifactFetchManager.java

```diff
@@ -0,0 +1,185 @@
+                        .collect(Collectors.toList());
+
+        if (artifacts.size() > 1) {
+            return new Result(null, artifacts);
+        }
+
+        if (artifacts.size() == 1) {
+            return new Result(artifacts.get(0), null);
+        }
+
+        // Should not happen.
+        throw new IllegalStateException("Corrupt artifact fetching state.");
+    }
+
+    /**
+     * Fetches the job jar and any additional artifact if the given list is not null or empty.
+     *
+     * @param jobUri URI of the job jar
+     * @param additionalUris URI(s) of any additional artifact to fetch
+     * @return result with the fetched artifacts
+     * @throws Exception
+     */
+    public Result fetchArtifacts(String jobUri, @Nullable List<String> additionalUris)
+            throws Exception {
+        checkArgument(jobUri != null && !jobUri.trim().isEmpty(), "The jobUri is required.");
+
+        ArtifactUtils.createMissingParents(baseDir);
+        File jobJar = fetchArtifact(jobUri);
+
+        List<File> additionalArtifacts =
+                additionalUris == null
+                        ? Collections.emptyList()
+                        : additionalUris.stream()
+                                .map(FunctionUtils.uncheckedFunction(this::fetchArtifact))
+                                .collect(Collectors.toList());
+
+        return new Result(jobJar, additionalArtifacts);
+    }
+
+    private File fetchArtifact(String uri) throws Exception {
+        URI resolvedUri = PackagedProgramUtils.resolveURI(uri);
+        File targetFile = new File(baseDir, FilenameUtils.getName(resolvedUri.getPath()));
+        if (targetFile.exists()) {
+            // Already fetched user artifacts are kept.
+            return targetFile;
+        }
+
+        return getFetcher(resolvedUri).fetch(uri, conf, baseDir);
+    }
+
+    @VisibleForTesting
+    ArtifactFetcher getFetcher(URI uri) {
+        if ("local".equals(uri.getScheme())) {
+            return localFetcher;
+        }
+
+        if ("http".equals(uri.getScheme()) || "https".equals(uri.getScheme())) {
+            return httpFetcher;
+        }
+
+        return fsFetcher;
+    }
+
+    /** Artifact fetch result with all fetched artifact(s). */
+    public static class Result {
+
+        private final File jobJar;
+        private final List<File> additionalJars;
+
+        private Result(@Nullable File jobJar, @Nullable List<File> additionalJars) {
+            this.jobJar = jobJar;
+            this.additionalJars = additionalJars == null ? Collections.emptyList() : additionalJars;
+        }
+
+        @Nullable
+        public File getJobJar() {
+            return jobJar;
+        }
+
+        public List<File> getAdditionalJars() {
+            return Collections.unmodifiableList(additionalJars);
+        }
+
+        @Nullable
+        public File getUserArtifactDir() {
+            return additionalJars.isEmpty() ? null : additionalJars.get(0).getParentFile();
```

Comment:
Correct! Although after I thought this through again, this highlights a bug in the current behavior, because the `LocalArtifactFetcher` at the moment does not copies the artifacts, only gets their local path, and in case we pass for example `local:///tmp/foo/foo.jar, local:///tmp/baz/baz.jar` only 1 of them will be loaded to the CP, cause the `DefaultPackagedProgramRetriever` won't know about the other dir.

2 possible solutions that come into my mind would be:
1. Do copy `local://` artifacts to the `ARTIFACT_BASE_DIR` as well.
2. Instead of passing the `ARTIFACT_BASE_DIR`, pass the list of artifacts to `DefaultPackagedProgramRetriever`.

I do not really like option 1. because of doubling the required disk space. Option 2 should not add much complexity anyways, I went with passing the directory because this way it was possible to reuse loading the contents as it almost the same as `userLibDir`.